HW3

JoshuaWachana opened the pull request:
Work in progress, review it later

hickamt commented on hw3/02-charts.js:
```diff
@@ -5,48 +5,128 @@ const backgroundColors = [
+  for (house in houseData) {
+    listOfHouses.push(house);
+    const key = document.createElement('div');
+    key.innerHTML = `<div class="key-container"><div class="key-color-box" style="background-color: ${backgroundColors[indexColor]}"></div><div class="key">${house}</div></div>`;
```
You can replace the template literal with the following. I have also added a post-increment for indexColor which will increment the same as your last line of `indexColor += 1`. Of course I am not sure about the style guide on this one.

```js
 for (house in houseData) {
    listOfHouses.push(house);
    const keyContainer = document.createElement("div");
    keyContainer.className = "key-container";

    const keyColorBox = document.createElement("div");
    keyColorBox.className = "key-color-box";
    keyColorBox.style.backgroundColor = backgroundColors[indexColor++];

    const key = document.createElement("div");
    key.className = "key";
    key.textContent = house;

    keyContainer.append(keyColorBox, key);
    chartLegend.append(keyContainer);

    numberOfMembersList.push(houseData[house]);
  }
```

JoshuaWachana replied:
You are right, creating elements like this is not ideal. I have refactored it as per your suggestion.

hickamt commented on hw3/02-charts.js:
```diff
@@ -5,48 +5,128 @@ const backgroundColors = [
+    const response = await fetch(url);
+    const data = await response.json();
+    calculateHouseMembers(data);
+  } catch (error) {}
```
You could handle that `catch (error) {}` with console.log(error)

JoshuaWachana replied:
I should not have left this empty, I guess I was in a rush and totally forgot about it. I have fixed this now, I decided to append an error message to the DOM incase something wrong happens with the API call. I though this would be more appropriate than just printing to the console since most casual users do not check the console for error messages.

hickamt commented on hw3/02-charts.js:
```diff
@@ -5,48 +5,128 @@ const backgroundColors = [
   const donutChart = document.querySelector('.donut-chart');
+  const chartLegend = document.querySelector('.chart-legend');
+
+  // const donutChart = document.getElementById('chart-1').getContext('2d');
```
Maybe, remove the comment here on line 37 and the `console.log(houseData)`

JoshuaWachana replied:
True, those two lines were unnecessary. I have removed them.

hickamt commented on hw3/02-charts.js:
```diff
@@ -5,48 +5,128 @@ const backgroundColors = [
     type: 'doughnut',
     data: {
-      labels: ['label', 'label', 'label', 'label'],
+      labels: [...listOfHouses],
```
ChartJS will handle `labels: listOfHouses` for your. You don't need to add the spread operator which will cut down on two array iterations.

JoshuaWachana replied:
This suggestion totally makes sense, I have implemented your suggestion.

hickamt commented on hw3/02-charts.js:
```diff
@@ -5,48 +5,128 @@ const backgroundColors = [
+          label: 'Members per House',
+          data: [...numberOfMembersList],
           backgroundColor: backgroundColors,
           borderColor: borderColors,
```
You can remove the `const borderColors` to remove another array copy iteration from `[...backgroundColors]` and just name `backgroundColors` as `const chartColors` and feed the object to both `backgroundColor` && `bordercolor`.

JoshuaWachana replied:
Thanks for the tip!

hickamt commented on hw3/ex03-react/public/index.html:
```diff
@@ -0,0 +1,43 @@
+    <title>React App</title>
+  </head>
+  <body>
+    <noscript>You need to enable JavaScript to run this app.</noscript>
```
WAVE is giving an alert that this `<noscript>` element is on each page. You may want to remove this.

JoshuaWachana replied:
Thanks, I have addressed this.

hickamt commented on hw3/ex03-react/src/chartColors.js:
```diff
@@ -0,0 +1,32 @@
+
+export default backgroundColors;
+
+export { borderColors };
```
Not a great practice to have a default and an export object in the same file. 
For config and other static objects just add:
`export const backgroundColors` and `export const borderColors` and remove the `export default .. && export { ... }` that you are currently using.

hickamt commented on hw3/ex03-react/src/components/Chart.js:
```diff
@@ -0,0 +1,61 @@
+import React from 'react';
+import { Chart as ChartJS } from 'chart.js/auto';
+import backgroundColors from '../chartColors';
```
I think I found out why you had two different export types in the /chartcolors.js file.
When importing two static objects from a file:
`
export const objectOne = []
export const objectTwo = []
`
Do not use default for one and try to export the other.

Then, when importing a non-default object you will use this import syntax:
`
import { backgroundColors, borderColors} from 'chart.js/auto';
`
And this will allow you to use them in the ChartJS datasets options:
`
{
backgroundColor: backgroundColors,
borderColor: borderColors,
}
`
I tried it and it works the same as you implemented in your 02-charts.js exercise.

JoshuaWachana replied:
Thanks for the tip!

hickamt reviewed (changes requested):
## Exercise 1

Code looks good. As the render is only once I guess using the `const func = function longerFunctionName() {}` isn't required.

STYLE:

The homework has a screen shot at 414px which shows two character cards in the frame. You currently show only one card.

## Exercise 2

This exercise does not look like your typical clean code. Guessing it was a busy week, midterms and all.

CODE:

Give the 'error' for try catch a console.error(error).

You can remove the [...ObjectArray] spread operator for `labels` and `data` as ChartJS expects an array. This will cut down on two array iterations.

I have included working code that removes the template literal and builds the div elements as you typically have in your code.

STYLE:

The chart looks great and I like the idea of moving the legend to another div. A downside of this is the loss of functionality that you would have with the legend on click ability to remove a section of the chart.

## Exercise 3

This application works as expected and fulfills the homework requirements.

CODE:

- Remove the `<noscript>` element from the /public/index.html
- Change your object exports for /chartColors.js. Just use `export const objectName` for both.
- /components/Chart.js: I left a note for you regarding changes that can be made to allow the use of backgroundColors and borderColors. This is directly related to how you are trying to export the static color objects from /chartColors.js.
- You might try copying the Bootstrap script into /public/index.html and using Bootstrap to create a styled navbar.

ACCESSIBILITY:

- /Home: you left the `<noscript> ... </noscript>` element in the HTML which is placed there by `create-react-app` I believe. Just remove it as WAVE is giving an Alert.

- NavLinks: The links have `Very low contrast` according to WAVE.

- /Search: The input is `Missing a form label` according to WAVE.

linh224 reviewed (commented):
Hi Joshua,
**Exercise 1:** The requirement for mobile view is to display 2 cards, I see you show 1 card. In my exercise I show 3 cards :))). I am trying to do as the screenshot. 

**Exercise 2**: Your chart appears great. You performed nicely at 414 px and 1024 px.

**Exercise 3:**
I think your README.md needs to be updated. That would be better to the original README.md
When I performed Wave testing, it reported 3 “Contrast Errors” and 1 Alerts “Noscript element”. There is 1 a 1 alert “Orphaned form label” in tab Search
Overall, this program functions exactly as expected. :)

JoshuaWachana commented:
> Hi Joshua, **Exercise 1:** You only display three cards at 1024 pixels, although the requirement calls for the display of four cards. The requirement for mobile view is to display 2 cards, I see you show 1 card. In my exercise I show 3 cards :))). I am trying to do as the screenshot.
> 
> **Exercise 2**: Your chart appears great. You performed nicely at 414 px and 1024 px.
> 
> **Exercise 3:** I think your README.md needs to be updated. That would be better to the original README.md When I performed Wave testing, it reported 3 “Contrast Errors” and 1 Alerts “Noscript element”. There is 1 a 1 alert “Orphaned form label” in tab Search Overall, this program functions exactly as expected. :)

Thanks for the feedback Linh! 

I have addressed some of the feedback and updated my PR.

JoshuaWachana commented:
> ## Exercise 1
> Code looks good. As the render is only once I guess using the `const func = function longerFunctionName() {}` isn't required.
> 
> STYLE:
> 
> The homework has a screen shot at 414px which shows two character cards in the frame. You currently show only one card.
> 
> ## Exercise 2
> This exercise does not look like your typical clean code. Guessing it was a busy week, midterms and all.
> 
> CODE:
> 
> Give the 'error' for try catch a console.error(error).
> 
> You can remove the [...ObjectArray] spread operator for `labels` and `data` as ChartJS expects an array. This will cut down on two array iterations.
> 
> I have included working code that removes the template literal and builds the div elements as you typically have in your code.
> 
> STYLE:
> 
> The chart looks great and I like the idea of moving the legend to another div. A downside of this is the loss of functionality that you would have with the legend on click ability to remove a section of the chart.
> 
> ## Exercise 3
> This application works as expected and fulfills the homework requirements.
> 
> CODE:
> 
> * Remove the `<noscript>` element from the /public/index.html
> * Change your object exports for /chartColors.js. Just use `export const objectName` for both.
> * /components/Chart.js: I left a note for you regarding changes that can be made to allow the use of backgroundColors and borderColors. This is directly related to how you are trying to export the static color objects from /chartColors.js.
> * You might try copying the Bootstrap script into /public/index.html and using Bootstrap to create a styled navbar.
> 
> ACCESSIBILITY:
> 
> * /Home: you left the `<noscript> ... </noscript>` element in the HTML which is placed there by `create-react-app` I believe. Just remove it as WAVE is giving an Alert.
> * NavLinks: The links have `Very low contrast` according to WAVE.
> * /Search: The input is `Missing a form label` according to WAVE.

Thanks for the feedback Todd!

I have addressed some of the feedback and updated my PR.
